include/lib/address_bitset.hpp

```diff
@@ -0,0 +1,61 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0. This product includes software
+// developed at Datadog (https://www.datadoghq.com/). Copyright 2021-Present
+// Datadog, Inc.
+#pragma once
+
+#include <atomic>
+#include <memory>
+#include <stdint.h>
+#include <string.h>
+
+namespace ddprof {
+class AddressBitset {
+  // Number of bits is the number of addresses we can store
+  // We have one address per individual bit).
+  // so lets say you have 1111, you can store 4 addresses.
+  // We hash the address to a number (to have an equal probability of using
+  // all bits). Then we use the mask to position this address in our bitset.
+  // Addr -> Hash -> Mask (to get useable bits) -> Position in the bitset
+  // Note: the hashing step might be bad for cache locality.
+public:
+  // Publish 1 Meg as default
+  constexpr static unsigned _k_default_max_addresses = 8 * 1024 * 1024;
+  AddressBitset(unsigned max_addresses = 0) { init(max_addresses); }
+  void init(unsigned max_addresses);
+  // returns true if the element was inserted
+  bool set(uintptr_t addr);
+  // returns true if the element was removed
+  bool unset(uintptr_t addr);
+  void clear();
+  int nb_addresses() const { return _nb_addresses; }
+
+private:
+  static constexpr unsigned _k_max_bits_ignored = 4;
+  unsigned _lower_bits_ignored;
+  // element type
+  using Elt_t = uint64_t;
+  constexpr static unsigned _nb_bits_per_elt = sizeof(Elt_t) * 8;
+  // 1 Meg divided in uint64's size
+  // The probability of collision is proportional to the number of elements
+  // already within the bitset
+  unsigned _nb_bits = {};
+  unsigned _k_nb_elements = {};
+  unsigned _nb_bits_mask = {};
+  // We can not use an actual bitset (for atomicity reasons)
+  std::unique_ptr<std::atomic<uint64_t>[]> _address_bitset;
+  std::atomic<int> _nb_addresses = 0;
+
+  // This is a kind of hash function
+  // We remove the lower bits (as the alignment constraints makes them useless)
+  // We fold the address
+  // Then we only keep the bits that matter for the order in the bitmap
+  uint32_t remove_lower_bits(uintptr_t h1) {
+    uint64_t intermediate = h1 >> _lower_bits_ignored;
+    uint32_t high = (uint32_t)(intermediate >> 32);
+    uint32_t low = (uint32_t)intermediate;
+    uint32_t res = high ^ low;
+    return res & _nb_bits_mask;
+  }
```

Comment:
I think you could get away with exactness here:
* you can strip the top 16 bits out anyway, so you only need to actually store 48 bits, on top of the up to 4 you can remove for alignment.
* if you only need to store `48 - alignment` bits, you can consider a paginated structure, where you have up to `2^(32 - alignment)` pages of 16 bit bitsets. This means you have an enormous array of null pointers, occasionally pointing to an 8KiB bitset whenever an address in the range [i, i + 2^16) has been encountered. 
* `2^(32 - alignment)` is going to be too large (at least 2GiB with the max alignment), but allocation patterns will be sparse, especially in the high order bits, so what about a 3 level structure with up to 2^16 pages for the top 16 bits, each (there will be very few of these populated) indexing paginated bitsets of `2^(16 - alignment)` pages of 16 bit bitsets. This pushes the alignment saving into the middle 12-16 bits which are likely to have higher cardinalities than the top 16. 
   * The top level costs 512 KiB, most of it will be null pointers
   * Each populated range in the top level will cost 32-512 KiB according to alignment (32 for 16 byte alignment, 512 for no alignment)
   * Each range of 2^16 addresses after alignment costs 8KiB for the leaf level bitset.
   * E.g. for an 8GiB address space and 8 byte alignment, you end up with a fixed 512KiB, 2 * 64 KiB for the 2 populated middle pages, and 8KiB for any range of 2^16 consecutive addresses where you sample a pointer.

Reply:
And you can trim the 512KiB a lot if you know the max address

Reply:
Thanks for exposing the math here.

Exactness can have value, especially for cases where the user puts a lower sampling interval and just wants to debug his program. I will delay trying this out as it will take me down a new cycles of experiments and measures.
On average the current approach is good to reduce memory impact. We do skew slightly the sampling intervals when a collision occurs.

Reply:
Added a ticket to internal board for now: https://datadoghq.atlassian.net/browse/PROF-8222

Reply:
A first draft around this topic: https://github.com/DataDog/ddprof/pull/307/files
Perhaps we need to have some quick access path. It looks like the current performances are not yet good enough.